chain-core/src/common/merkle_tree.rs

```diff
@@ -353,39 +361,17 @@ impl<T> MerkleTree<T> {
     }
 }
 
-/// A node can either be a leaf or intermediate node in a merkle tree
-#[derive(Debug)]
-pub enum NodeType {
-    Leaf,
-    Intermediate,
-}
-
-fn combine(left: &H256, right: &H256) -> H512 {
-    let mut hash = [0; HASH_SIZE_256 * 2];
-    hash[0..HASH_SIZE_256].copy_from_slice(left);
-    hash[HASH_SIZE_256..(HASH_SIZE_256 * 2)].copy_from_slice(right);
-
-    hash
-}
-
-pub fn hash<T: AsRef<[u8]>>(value: T, node_type: NodeType) -> H256 {
-    match node_type {
-        NodeType::Leaf => hash256::<Blake2s>(&prefix_with_byte(value, 0)),
-        NodeType::Intermediate => hash256::<Blake2s>(&prefix_with_byte(value, 1)),
```

Comment:
Why is this removed? This was here to prevent [second preimage attack](https://en.wikipedia.org/wiki/Merkle_tree#Second_preimage_attack).

Reply:
Will `keyed_hash` prevent this attack?

Reply:
If I understand it correctly, that attack is due to a lack of domain separation -- e.g. in that example:
- H0 and H1 are intermediate nodes
- or one can forge an equivalent where H0 and H1 will be "leafs" (with data hash 0-0 + hash 0-1 and hash 1-0 + hash 1-1)

Byte-prefixing fixes that, as e.g. hash(0x00 + hash 0-0 + hash 0-1) != hash(0x01 + hash 0-0 + hash 0-1) (plus those intermediate hashes hash 0-0 + hash 0-1 are different).

With a keyed hash, it seems similar, as hash_leaf(hash 0-0 + hash 0-1) != hash_node(hash 0-0 + hash 0-1) ?

Or does this open a way for other forgery?

Reply:
I can put it back, as it was

Reply:
I don't know. My intuition says it should be okay with `keyed_hash`, but we should make sure.

Reply:
switched it back -- while reading the code, I discovered one flaw: https://github.com/crypto-com/chain/issues/1329